dev, 3.0: Add tkctl user guide (#1354)
pingcap/docs: dev/reference/tools/tkctl.md

---
title: TiDB Kubernetes Control User Guide
category: reference
---

# TiDB Kubernetes Control User Guide

TiDB Kubernetes Control(`tkctl`) is a command line utility for TiDB operators to operate and diagnose their TiDB clusters in Kubernetes.

## Installation

You can download the pre-built binary or build `tkctl` from source:

### Download the Latest Pre-built Binary

- [MacOS](http://download.pingcap.org/tkctl-darwin-amd64-latest.tgz)
- [Linux](http://download.pingcap.org/tkctl-linux-amd64-latest.tgz)
- [Windows](http://download.pingcap.org/tkctl-windows-amd64-latest.tgz)

### Build from Source

{{< copyable "shell-regular" >}}

```shell

> git clone https://github.com/pingcap/tidb-operator.git && \

Reviewer: Add `--depth=1`?

```shell
tkctl version
```

If you see the version of tkctl tool and version of TiDB operator installed in target cluster or "No TiDB Controller Manager found, please install one first.", `tkctl` is correctly configured to access your cluster.

## Commands

### tkctl version

This command used to show the version of **tkctl** and **tidb-operator** installed in target cluster.

Example:

{{< copyable "shell-regular" >}}

```shell
tkctl version
```

```
Client Version: v1.0.0-beta.1-p2-93-g6598b4d3e75705-dirty
TiDB Controller Manager Version: pingcap/tidb-operator:latest
TiDB Scheduler Version: pingcap/tidb-operator:latest
```

### tkctl list

This command used to list all tidb clusters installed.

| Flags | Shorthand | Description |
| ----- | --------- | ----------- |
| --all-namespaces | -A | search all namespaces |
| --output | -o | output format, one of [default,json,yaml], the default format is `default` |

Example:

{{< copyable "shell-regular" >}}

```shell
tkctl list -A
```

```
NAMESPACE NAME           PD    TIKV   TIDB   AGE
foo       demo-cluster   3/3   3/3    2/2    11m
bar       demo-cluster   3/3   3/3    1/2    11m
```

### tkctl use

> This command used to specify the current TiDB cluster to use, the other commands could omit `--tidbcluster` option and defaults to select current TiDB cluster if there is a current TiDB cluster set.

Reviewer 2: ```suggestion
This command is used to specify the TiDB cluster that the current `tkctl` command operates on. After you specify a TiDB cluster by using this command, all commands that operates on a cluster will automatically select this cluster so the `--tidbcluster` option can be omitted.
```

> Example:

Reviewer 2: ```suggestion
For example:
```

{{< copyable "shell-regular" >}}

```shell
tkctl use --namespace=foo demo-cluster
```

```
Tidb cluster switched to foo/demo-cluster
```

### tkctl info

> This command used to get the information of TiDB cluster, the current TiDB cluster will be used if exists.

Reviewer 2: ```suggestion
This command is used to display information about the TiDB cluster.
```

> | Flags | Shorthand | Description |

Reviewer 2: ```suggestion
| Flag | Abbreviation | Description |
```

| ----- | --------- | ----------- |

> | --tidb-cluster | -t | select the tidb cluster, default to current TiDB cluster |

Reviewer 2: ```suggestion
| --tidb-cluster | -t | Specify a TiDB cluster; default to the TiDB cluster that is being used |
```

> Example:

Reviewer 2: ```suggestion
For example:
```

{{< copyable "shell-regular" >}}

```shell
tkctl info
```

```
Name:               demo-cluster
Namespace:          foo
CreationTimestamp:  2019-04-17 17:33:41 +0800 CST
Overview:
         Phase    Ready  Desired  CPU    Memory  Storage  Version
         -----    -----  -------  ---    ------  -------  -------
  PD:    Normal   3      3        200m   1Gi     1Gi      pingcap/pd:v3.0.0-rc.1
  TiKV:  Normal   3      3        1000m  2Gi     10Gi     pingcap/tikv:v3.0.0-rc.1
  TiDB   Upgrade  1      2        500m   1Gi              pingcap/tidb:v3.0.0-rc.1
Endpoints(NodePort):
  - 172.16.4.158:31441
  - 172.16.4.155:31441
```

### tkctl get [component]

> This is a group of commands used to get the details of TiDB cluster componentes, the current TiDB cluster will be used if exists.

Reviewer 2: ```suggestion
This is a group of commands that are used to get the details of TiDB cluster components.
```

> Available components: `pd`, `tikv`, `tidb`, `volume`, `all`(query all components)

Reviewer 2: ```suggestion
You can query the following components: `pd`, `tikv`, `tidb`, `volume` and `all` (to query all components).
```

> | Flags | Shorthand | Description |

Reviewer 2: ```suggestion
| Flag | Abbreviation | Description |
```
Update this table accordingly as commented above.

| ----- | --------- | ----------- |
| --tidb-cluster | -t | select the tidb cluster, default to current TiDB cluster |
| --output | -o | output format, one of [default,json,yaml], the default format is `default` |

> Example:

Reviewer 2: ```suggestion
For example:
```

```

```
NAME                  READY   STATUS    MEMORY          CPU   RESTARTS   AGE     NODE
demo-cluster-tikv-0   2/2     Running   2098Mi/4196Mi   2/2   0          3m19s   172.16.4.155
demo-cluster-tikv-1   2/2     Running   2098Mi/4196Mi   2/2   0          4m8s    172.16.4.160
demo-cluster-tikv-2   2/2     Running   2098Mi/4196Mi   2/2   0          4m45s   172.16.4.157
```

{{< copyable "shell-regular" >}}

```shell
tkctl get volume
```

```
VOLUME              CLAIM                      STATUS   CAPACITY   NODE           LOCAL
local-pv-d5dad2cf   tikv-demo-cluster-tikv-0   Bound    1476Gi     172.16.4.155   /mnt/disks/local-pv56
local-pv-5ade8580   tikv-demo-cluster-tikv-1   Bound    1476Gi     172.16.4.160   /mnt/disks/local-pv33
local-pv-ed2ffe50   tikv-demo-cluster-tikv-2   Bound    1476Gi     172.16.4.157   /mnt/disks/local-pv13
local-pv-74ee0364   pd-demo-cluster-pd-0       Bound    1476Gi     172.16.4.155   /mnt/disks/local-pv46
local-pv-842034e6   pd-demo-cluster-pd-1       Bound    1476Gi     172.16.4.158   /mnt/disks/local-pv74
local-pv-e54c122a   pd-demo-cluster-pd-2       Bound    1476Gi     172.16.4.156   /mnt/disks/local-pv72
```

### tkctl debug [pod_name]

> This command used to diagnose the Pods of TiDB cluster. It launches a debug container for you which has the necessary troubleshooting tools installed.

Reviewer 2: ```suggestion
This command is used to diagnose the Pods in a TiDB cluster. It launches a debug container for you which has the necessary troubleshooting tools installed.
```

>
> The default image of the debug container contains various troubleshooting tools, so the image size is relatively large. If you only need `pd-ctl` and `tidb-ctl`, you can use the `tidb-control` image by specifying `--image=pingcap/tidb-control:latest` command line option.

Example:

{{< copyable "shell-regular" >}}

```shell
tkctl debug demo-cluster-tikv-0
```

{{< copyable "shell-regular" >}}

```shell
ps -ef
```

Using tools like `GDB` and `perf` in the debug container requires special operations because of the difference in root filesystems of the target container and the debug container.

#### GDB

When using GDB, make sure you point the `program` option to the binary in the **target container**. Additionally, if you use images other than `tidb-debug` as debug container or the `pid` of the target process is not 1, you have to configure the location of dynamic libraries via the `set sysroot` command, steps are as follows:

{{< copyable "shell-regular" >}}

```shell
tkctl debug demo-cluster-tikv-0
```

{{< copyable "shell-regular" >}}

```shell
gdb /proc/${pid:-1}/root/tikv-server 1
```

{{< copyable "shell-regular" >}}

```shell
(gdb) set sysroot /proc/${pid}/root/
```

> The `.gdbinit` configured in the `tidb-debug` image will set `sysroot` to `/proc/1/root/` automatically. So, you can omit this following step if you are using the `tidb-debug` image and the `pid` of the target process 1.

Reviewer: "this following step" should be "the above step"?
